include/sound/sof/ipc4/header.h

```diff
@@ -0,0 +1,338 @@
+enum sof_ipc4_msg_dir {
+	SOF_IPC4_MSG_REQUEST = 0,
+	SOF_IPC4_MSG_REPLY,
+};
+
+enum sof_ipc4_pipe_line_state {
+	SOF_IPC4_PIPE_INVALID_STATE = 0,
+	SOF_IPC4_PIPE_UNINITIALIZED,
+	SOF_IPC4_PIPE_RESET,
+	SOF_IPC4_PIPE_PAUSED,
+	SOF_IPC4_PIPE_RUNNING,
+	SOF_IPC4_PIPE_EOS
+};
+
+/* global common ipc msg */
+#define SOF_IPC4_GLB_MSG_TARGET_SHIFT		30
+#define SOF_IPC4_GLB_MSG_TARGET_MASK		BIT(30)
+#define SOF_IPC4_GLB_MSG_TARGET(x)		((x) << SOF_IPC4_GLB_MSG_TARGET_SHIFT)
+
+#define SOF_IPC4_GLB_MSG_DIR_SHIFT		29
+#define SOF_IPC4_GLB_MSG_DIR_MASK		BIT(29)
+#define SOF_IPC4_GLB_MSG_DIR(x)			((x) << SOF_IPC4_GLB_MSG_DIR_SHIFT)
+
+#define SOF_IPC4_GLB_MSG_TYPE_SHIFT		24
+#define SOF_IPC4_GLB_MSG_TYPE_MASK		GENMASK(28, 24)
+#define SOF_IPC4_GLB_MSG_TYPE(x)		((x) << SOF_IPC4_GLB_MSG_TYPE_SHIFT)
+
+/* pipeline creation ipc msg */
+#define SOF_IPC4_GLB_PIPE_INSTANCE_SHIFT	16
+#define SOF_IPC4_GLB_PIPE_INSTANCE_MASK		GENMASK(23, 16)
+#define SOF_IPC4_GLB_PIPE_INSTANCE_ID(x)	((x) << SOF_IPC4_GLB_PIPE_INSTANCE_SHIFT)
+
+#define SOF_IPC4_GLB_PIPE_PRIORITY_SHIFT	11
+#define SOF_IPC4_GLB_PIPE_PRIORITY_MASK		GENMASK(15, 11)
+#define SOF_IPC4_GLB_PIPE_PRIORITY(x)		((x) << SOF_IPC4_GLB_PIPE_PRIORITY_SHIFT)
+
+#define SOF_IPC4_GLB_PIPE_MEM_SIZE_SHIFT	0
+#define SOF_IPC4_GLB_PIPE_MEM_SIZE_MASK		GENMASK(10, 0)
+#define SOF_IPC4_GLB_PIPE_MEM_SIZE(x)		((x) << SOF_IPC4_GLB_PIPE_MEM_SIZE_SHIFT)
+
+#define SOF_IPC4_GL_PIPE_EXT_LP_SHIFT		0
+#define SOF_IPC4_GL_PIPE_EXT_LP_MASK		BIT(0)
+#define SOF_IPC4_GL_PIPE_EXT_LP(x)		((x) << SOF_IPC4_GL_PIPE_EXT_LP_SHIFT)
+
+/* pipeline set state ipc msg */
+#define SOF_IPC4_GL_PIPE_STATE_TYPE_SHIFT	24
+#define SOF_IPC4_GL_PIPE_STATE_TYPE_MASK	GENMASK(28, 24)
+#define SOF_IPC4_GL_PIPE_STATE_TYPE(x)		((x) << SOF_IPC4_GL_PIPE_STATE_TYPE_SHIFT)
+
+#define SOF_IPC4_GL_PIPE_STATE_ID_SHIFT		16
+#define SOF_IPC4_GL_PIPE_STATE_ID_MASK		GENMASK(23, 16)
+#define SOF_IPC4_GL_PIPE_STATE_ID(x)		((x) << SOF_IPC4_GL_PIPE_STATE_ID_SHIFT)
+
+#define SOF_IPC4_GL_PIPE_STATE_SHIFT		0
+#define SOF_IPC4_GL_PIPE_STATE_MASK		GENMASK(15, 0)
+#define SOF_IPC4_GL_PIPE_STATE(x)		((x) << SOF_IPC4_GL_PIPE_STATE_SHIFT)
+
+enum sof_ipc4_sampling_frequency {
+	SOF_IPC4_FS_8000HZ	= 8000,
+	SOF_IPC4_FS_11025HZ	= 11025,
+	SOF_IPC4_FS_12000HZ	= 12000,
+	SOF_IPC4_FS_16000HZ	= 16000,
+	SOF_IPC4_FS_18900HZ	= 18900,
+	SOF_IPC4_FS_22050HZ	= 22050,
+	SOF_IPC4_FS_24000HZ	= 24000,
+	SOF_IPC4_FS_32000HZ	= 32000,
+	SOF_IPC4_FS_37800HZ	= 37800,
+	SOF_IPC4_FS_44100HZ	= 44100,
+	SOF_IPC4_FS_48000HZ	= 48000, /* Default. */
+	SOF_IPC4_FS_64000HZ	= 64000,
+	SOF_IPC4_FS_88200HZ	= 88200,
+	SOF_IPC4_FS_96000HZ	= 96000,
+	SOF_IPC4_FS_176400HZ	= 176400,
+	SOF_IPC4_FS_192000HZ	= 192000,
+};
+
+enum sof_ipc4_bit_depth {
+	SOF_IPC4_DEPTH_8BIT	= 8, /* 8 bits depth */
+	SOF_IPC4_DEPTH_16BIT	= 16, /* 16 bits depth */
+	SOF_IPC4_DEPTH_24BIT	= 24, /* 24 bits depth - Default */
+	SOF_IPC4_DEPTH_32BIT	= 32, /* 32 bits depth */
+	SOF_IPC4_DEPTH_64BIT	= 64, /* 64 bits depth */
```

Comment:
Same here, why we need  enum for these? How many bits are there to store it? 7? 8?

Reply:
In C language, an enum is guaranteed to be of size of an int. There is a compile time option (-fshort-enums) to make it as short (This is mainly useful in case the values are not more than 64K). There is no compile time option to increase its size to 64 bit.

Reply:
some compilers have options to break the standard and use the smallest type that will contain all values.

Example with GCC (documentation in the GCC Manual):

enum ord {
    FIRST = 1,
    SECOND,
    THIRD
} __attribute__ ((__packed__));
STATIC_ASSERT( sizeof(enum ord) == 1 )

Reply:
Sure, and there is the `-fshort-enums`  and on 16bit processors int is 16bit.
I know that we will not using either of these, but Seeing enum in a struct which is used between two different processors as IPC message does not sound right.

Reply:
Ack, this doesn't seem to add anything. We could set the value directly and keep existing FW happy.